From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Sat, 9 May 2026 10:32:06 +0800
Subject: [PATCH 21/21] input: misc: macsmc: Force restart instead of poweroff
 on T8015

In iOS, asserting force reset causes the device to restart, not poweroff.
Match that behavior here.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 drivers/input/misc/macsmc-input.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/input/misc/macsmc-input.c b/drivers/input/misc/macsmc-input.c
index a2de77d95..b3157f1f9 100644
--- a/drivers/input/misc/macsmc-input.c
+++ b/drivers/input/misc/macsmc-input.c
@@ -21,6 +21,7 @@
  * @input: Allocated input_dev; devres managed
  * @nb: Notifier block used for incoming events from SMC (e.g. button pressed down)
  * @wakeup_mode: Set to true when system is suspended and power button events should wake it
+ * @force_shutdown: True if force reset should shutdown, false if system should restart
  */
 struct macsmc_input {
 	struct device *dev;
@@ -28,6 +29,7 @@ struct macsmc_input {
 	struct input_dev *input;
 	struct notifier_block nb;
 	bool wakeup_mode;
+	bool force_shutdown;
 };
 
 #define SMC_EV_BTN 0x7201
@@ -80,8 +82,9 @@ static void macsmc_input_event_button(struct macsmc_input *smcin, unsigned long
 		 * flushed. macOS actually does this by panicing (!)...
 		 */
 		if (state) {
-			dev_crit(smcin->dev, "Triggering forced shutdown!\n");
-			if (kernel_can_power_off())
+			dev_crit(smcin->dev, "Triggering forced %s!\n",
+				 smcin->force_shutdown ? "shutdown" : "restart");
+			if (smcin->force_shutdown && kernel_can_power_off())
 				kernel_power_off();
 			else /* Missing macsmc-reboot driver? */
 				kernel_restart("SMC power button triggered restart");
@@ -154,6 +157,9 @@ static int macsmc_input_probe(struct platform_device *pdev)
 	if (!smcin->input)
 		return -ENOMEM;
 
+	if (!of_device_is_compatible(pdev->dev.parent->of_node, "apple,t8015-smc"))
+		smcin->force_shutdown = true;
+
 	smcin->input->phys = "macsmc-input (0)";
 	smcin->input->name = "Apple SMC power/lid events";
 
